### PR TITLE
Entity set aggregation

### PR DESCRIPTION
**Pull request by athoscouto**
### Issues
*This pull request fixes issue  #893.*  

### Description
*This PR adds capacity for WebAPI to handle `AggregateExpression ` of the type `EntitySetAggregate` and includes E2E tests cases for that feature. In order to do it I had to modify mostly the AggregationBinder, but changes to `ODataResourceSerializer` and `AggregationDynamicTypeProvider` were also needed, among other minor changes.*

### Checklist (Uncheck if it is not completed)
- [ ] *Unit test cases added.*
- [x] *E2E test cases added.*
- [ ] *Build and test with one-click build and test script passed.*

### CODE REVIEW

**Comment by robward-ms** [comment]
Can you add a description?  Also, this is a lot of commit history to add to OData, can you clean it up a bit?  I'd expect 1 or 2 commits, not 12.

**Comment by robward-ms** [comment]
This change requires an ODL change, it's PR is here: https://github.com/OData/odata.net/pull/757/.
This is blocked until that PR is accepted.

**Comment by xuzhg** [comment]
@kosinsky Would you please also rebase to the latest codes.  https://github.com/OData/WebApi/pull/1508 was merged several mins ago.

**Comment by kosinsky** [comment]
@xuzhg I've just merged with master and cleaned up.

File: src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs [hunk 1]
```diff
@@ -233,12 +233,14 @@ private void WriteDeltaResource(object graph, ODataWriter writer, ODataSerialize
             Contract.Assert(dynamicTypeProperties != null);
 
             var properties = new List<ODataProperty>();
-            var dynamicObject = graph as DynamicTypeWrapper;
+            var dynamicObject = graph as DynamicTypeWrapper ?? (graph as IEnumerable<DynamicTypeWrapper>).Single();
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs]: >graph as IEnumerable<DynamicTypeWrapper> [](start = 64, length = 40)

may it be "null"?

File: src/Microsoft.AspNet.OData.Shared/Query/Expressions/AggregationBinder.cs [hunk 2]
```diff
@@ -233,37 +237,131 @@ private IQueryable BindSelect(IQueryable grouping)
...
+            foreach (var aggExpression in expression.Children)
+            {
+                properties.Add(new NamedPropertyExpression(Expression.Constant(aggExpression.Alias), CreateAggregationExpression(innerAccum, aggExpression, selectedElementType)));
+            }
+
+            var nestedResultType = typeof(EntitySetAggregationWrapper);
+            var wrapperProperty = nestedResultType.GetProperty("Container");
+            wrapperTypeMemberAssignments.Add(Expression.Bind(wrapperProperty, AggregationPropertyContainer.CreateNextNamedPropertyContainer(properties)));
+
+            var initilizedMember =
```
Comment (AlanWong-MS) [inline src/Microsoft.AspNet.OData.Shared/Query/Expressions/AggregationBinder.cs]: nit: typo initilizedMember -> initializedMember

File: src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs [hunk 3]
```diff
@@ -233,12 +233,14 @@ private void WriteDeltaResource(object graph, ODataWriter writer, ODataSerialize
...
             if (dynamicObject != null)
             {
                 foreach (var prop in dynamicObject.Values)
                 {
-                    if (prop.Value != null && EdmLibHelpers.IsDynamicTypeWrapper(prop.Value.GetType()))
+                    if (prop.Value != null
+                        //&& EdmLibHelpers.IsDynamicTypeWrapper(prop.Value.GetType())
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs]: >//&& EdmLibHelpers.IsDynamicTypeWrapper(prop.Value.GetType()) [](start = 24, length = 61)

remove

File: src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs (hunk 1 above)
Comment (AlanWong-MS) [inline src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs]: What is the expected behavior if this is null?

File: src/Microsoft.AspNet.OData.Shared/ExpressionHelperMethods.cs [hunk 4]
```diff
@@ -23,6 +23,7 @@ internal class ExpressionHelperMethods
         private static MethodInfo _thenByDescendingMethod = GenericMethodOf(_ => Queryable.ThenByDescending<int, int>(default(IOrderedQueryable<int>), default(Expression<Func<int, int>>)));
         private static MethodInfo _enumerableThenByDescendingMethod = GenericMethodOf(_ => Enumerable.ThenByDescending<int, int>(default(IOrderedEnumerable<int>), default(Func<int, int>)));
         private static MethodInfo _countMethod = GenericMethodOf(_ => Queryable.LongCount<int>(default(IQueryable<int>)));
+        private static MethodInfo _enumerableGroupByMethod = GenericMethodOf(_ => Enumerable.GroupBy<int, int>(default(IQueryable<int>), default(Func<int, int>)));
```
[the remark this answers is not in the document]
Comment (kosinsky) [reply src/Microsoft.AspNet.OData.Shared/ExpressionHelperMethods.cs]: _enumerableGroupByMethod is used in EnumerableGroupByGeneric property that in use too.

File: src/Microsoft.AspNet.OData.Shared/Query/Expressions/AggregationBinder.cs [hunk 5]
```diff
@@ -233,37 +237,131 @@ private IQueryable BindSelect(IQueryable grouping)
...
+        private Expression CreateAggregationExpression(ParameterExpression accum, AggregateExpressionBase expression, Type baseType)
+        {
+            switch (expression.AggregateKind)
+            {
+                case AggregateExpressionKind.PropertyAggregate:
+                    return CreatePropertyAggregateExpression(accum, expression as AggregateExpression, baseType);
+                case AggregateExpressionKind.EntitySetAggregate:
+                    return CreateEntitySetAggregateExpression(accum, expression as EntitySetAggregateExpression, baseType);
+                default:
+                    throw new ODataException("This type of aggregation is not supported.");
```
Comment (AlanWong-MS) [inline src/Microsoft.AspNet.OData.Shared/Query/Expressions/AggregationBinder.cs]: We have a section for exception messages:
src/Microsoft.AspNet.OData.Shared/Common/SRResources.Designer.cs
src/Microsoft.AspNet.OData.Shared/Common/SRResources.resx

Could we also include the AggregateKind in the exception message?

Comment (xuzhg) [reply src/Microsoft.AspNet.OData.Shared/Query/Expressions/AggregationBinder.cs]: also include the type information in the error message.

---
In reply to: [198944947](https://github.com/OData/webapi/pull/906#discussion_r198944947) [](ancestors = 198944947)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/EntitySetAggregation/EntitySetAggregationDataModel.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data.Entity;
+
+namespace Microsoft.Test.E2E.AspNet.OData.EntitySetAggregation
+{
+    public class AggregationContext : DbContext
+    {
+        public static string ConnectionString =
+            @"Data Source=(LocalDb)\MSSQLLocalDB;Integrated Security=True;Initial Catalog=EntitySetAggregationTest1";
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/EntitySetAggregation/EntitySetAggregationDataModel.cs]: >EntitySetAggregationTest1 [](start = 90, length = 25)

Can we rename it?

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/EntitySetAggregation/EntitySetAggregationTests.cs [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            var result = await response.Content.ReadAsObject<JObject>();
+            var value = result["value"];
+            var orders = value.First["Orders"];
+            var totalPrice = orders.First["TotalPrice"].ToObject<int>();
+            var totalId = orders.First["TotalId"].ToObject<int>();
+
+            // OBS: DB uses sequential ID
+            // Each Customer has 2 orders that cost 25*CustomerId and 75*CustomerId
+            Assert.Equal(1*(25+75)+2*(25+75)+3*(25+75), totalPrice);
+            // Sum of the 6 Orders IDs, ignoreFunctionCalls
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/EntitySetAggregation/EntitySetAggregationTests.cs]: >ignoreFunctionCalls [](start = 40, length = 19)

ignoreFunctionCalls ? Maybe remove

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/EntitySetAggregation/EntitySetAggregationTests.cs [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var result = await response.Content.ReadAsObject<JObject>();
+            var value = result["value"];
+            var orders = value.First["Orders"];
+            var TotalPrice = orders.First["TotalPrice"].ToObject<int>();
+
+            Assert.Equal(1*(25+75)+2*(25+75)+3*(25+75), TotalPrice);
+        }
+
+
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/EntitySetAggregation/EntitySetAggregationTests.cs]: remove the extra while line

**Comment by xuzhg** [comment]
@kosinsky Thank for your contribution. It looks good to me except you can fix some small er comments. Thanks. 
@mikepizzo

**Comment by AlanWong-MS** [comment]
A couple of caught errors in the build:

```
2018-06-28T19:53:29.0938759Z ##[error]src\Microsoft.AspNet.OData.Shared\Query\Expressions\AggregationBinder.cs(249,95): Error : SA1001: Invalid spacing around the comma.
2018-06-28T19:53:29.0938759Z src\Microsoft.AspNet.OData.Shared\Query\Expressions\AggregationBinder.cs(249,95,249,95): error : SA1001: Invalid spacing around the comma. src\Microsoft.AspNet.OData\Microsoft.AspNet.OData.csproj]
2018-06-28T19:53:30.2189156Z ##[error]src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSerializer.cs(236,1): Error : SA0102: A syntax error has been discovered in file src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSerializer.cs on line 236.
```

**Comment by kosinsky** [comment]
I fixed these syntax errors. Looks like  WebApiOData.Performance.msbuild uses different compiler settings than other projects and doesn't support C# 6.0 features like ?.

**Comment by AlanWong-MS** [comment]
Looks good, thanks so much for the PR!

**Pull request merged**